### PR TITLE
WIP: ocgis

### PR DESCRIPTION
**Pull request by ocefpaf**
Ping @bekozi.

~~PS: I believe we have to tweak the `gdal` recipe.~~

~~Edit: This PR needs #158~~

### CODE REVIEW

**Comment by pelson** [comment]
Seems like there are a few issues with this one.

**Comment by ocefpaf** [comment]
I need to get back to this one at a later time. I will close it for now to reduce the noise here.

**Comment by pelson** [comment]
> I need to get back to this one at a later time. I will close it for now to reduce the noise here.

It's fine. I suspect this repo will never be free of noise - that is the nature of lots of timezones and day jobs. :smile:

**Comment by pelson** [comment]
> Selectors are suggested to take a <space><space>#<space>[<selector>] form. 

really is ugly trying to render the fact that it should be a leading double space "  # [selector]"

**Comment by ocefpaf** [comment]
:heart: the new label!

**Comment by pelson** [comment]
> :heart: the new label!

Hehe, that was me trying to get the linting service to recognise your bad selector :stuck_out_tongue_winking_eye:

**Comment by ocefpaf** [comment]
Ping @bekozi . (Re-building from that branch.)

**Comment by bekozi** [comment]
@ocefpaf, the failures on osx and windows are both related to ``gdal`` linking. For osx, the ``gdal`` shared library cannot be found by ``fiona``:

```
===== testing package: ocgis-1.3.0-py27_0 =====
import: 'ocgis'
Traceback (most recent call last):
  File "/Users/travis/miniconda/conda-bld/test-tmp_dir/run_test.py", line 25, in <module>
    import ocgis
  File "/Users/travis/miniconda/envs/_test/lib/python2.7/site-packages/ocgis/__init__.py", line 1, in <module>
    from ocgis.util.environment import env
  File "/Users/travis/miniconda/envs/_test/lib/python2.7/site-packages/ocgis/util/environment.py", line 10, in <module>
    from ocgis.util.helpers import get_iter
  File "/Users/travis/miniconda/envs/_test/lib/python2.7/site-packages/ocgis/util/helpers.py", line 10, in <module>
    import fiona
  File "/Users/travis/miniconda/envs/_test/lib/python2.7/site-packages/fiona/__init__.py", line 72, in <module>
    from fiona.collection import Collection, BytesCollection, vsi_path
  File "/Users/travis/miniconda/envs/_test/lib/python2.7/site-packages/fiona/collection.py", line 7, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/Users/travis/miniconda/envs/_test/lib/python2.7/site-packages/fiona/ogrext.so, 2): Library not loaded: /usr/local/lib/libgeotiff.2.dylib
  Referenced from: /Users/travis/miniconda/envs/_test/lib/libgdal.1.dylib
  Reason: image not found
TESTS FAILED: ocgis-1.3.0-py27_0
```

For windows, the issue is ``GDAL_DATA`` again. However, this looks like something I may be available to fix. Do you think the environment variable is appropriately set on the windows builds? I remember you saying that is was most likely fixed.

**Comment by ocefpaf** [comment]
> @ocefpaf, the failures on osx and windows are both related to gdal linking. For osx, the gdal shared library cannot be found by fiona:

I just rebased this PR because we are removing homebrew from Travis-CI now (and I forgot to do that here). That should fix gdal/fiona issues.

> Do you think the environment variable is appropriately set on the windows builds? I remember you saying that is was most likely fixed.

I need to check that again. I will get back to you.

**Comment by ocefpaf** [comment]
The failures have nothing to do with this PR. I am preparing a new `gdal` that should fix this.

**Comment by ocefpaf** [comment]
Ping @bekozi. Things are green on OS X and Linux now. The Windows failure is the `GDAL` folder issue.

[Here](https://github.com/conda-forge/gdal-feedstock/blob/master/recipe/bld.bat#L85-L94) is how we call the activate script and [here](https://github.com/conda-forge/gdal-feedstock/blob/master/recipe/scripts/activate.bat) is the script. Let me know if you can see any obvious errors and/or if that could be fixed in `ocgis`.

**Comment by bekozi** [comment]
Thanks @ocefpaf. I'll take a look.

File: recipes/ocgis/run_test.py [hunk 1]
```diff
@@ -1,3 +1,13 @@
+import os
 from ocgis.test import run_simple
 
+# Set GDAL_DATA. This is done normally done by the activate script,
+# but this doesn't happen in the testing environment.
+if 'LIBRARY_PREFIX' in os.environ:  # Windows.
+    gdalData = os.path.join(os.environ['LIBRARY_PREFIX'], 'share', 'gdal')
+else:  # Linux/OS X.
+    gdalData = os.path.join(os.environ['PREFIX'], 'share', 'gdal')
```
Comment (ocefpaf) [inline recipes/ocgis/run_test.py]: @bekozi this is the trick we do in `gdal`. In theory that var will be available in normal envs, not sure why it does not work on the `_test` env. Can `ocgis` use that var?

Comment (bekozi) [inline recipes/ocgis/run_test.py]: @ocefpaf, this should be usable. I can add it to the bag of tricks used to find the GDAL_DATA directory. Thanks for the snippet. I'll do some testing on my end.

**Comment by bekozi** [comment]
@ocefpaf, we can retry these builds. Hopefully for the last time. :smile:  I made a change to the path search for windows and was able to build successfully on linux and windows locally. You can revert ``run_test.py`` to:

```python
from ocgis.test import run_simple

run_simple()
```

The ``GDAL_DATA`` path moved with the new build, so I added another search option. The reason it wasn't working with your test file changes was that the environment variable was set below the ``ocgis`` import. However, we should be able to avoid this now!

**Comment by bekozi** [comment]
We'll need to wait for the bug fix version before merging of course...

File: recipes/ocgis/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
+{% set version = "1.3.0" %}
+
+package:
+    name: ocgis
+    version: {{ version }}
+
+source:
+    git_url: https://github.com/NCPP/ocgis.git
+    #git_tag: v{{ version }}
+    git_tag: i415-conda-forge
```
Comment (ocefpaf) [inline recipes/ocgis/meta.yaml]: I re-started the CIs. If that works do you plan on a new release or do you want to make the conda package using this branch for a while? (That is fine if you have reasons to do so.)

Comment (bekozi) [inline recipes/ocgis/meta.yaml]: There are a couple other bugs so I don't mind tagging one. Will take a day or so.

Comment (ocefpaf) [inline recipes/ocgis/meta.yaml]: No rush. We can wait for good software.

**Comment by ocefpaf** [comment]
And we have green! :tada:

**Comment by jakirkham** [comment]
So, I guess we are waiting for a new release. Correct?

**Comment by bekozi** [comment]
Excellent! @jakirkham, yes, I'll tag one in the near-term and report back here when it's ready.

File: recipes/ocgis/bld.bat [hunk 3]
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install --single-version-externally-managed --record record.txt
```
Comment (pelson) [inline recipes/ocgis/bld.bat]: We don't need this file anymore 😄

Comment (ocefpaf) [inline recipes/ocgis/bld.bat]: Happy to remove it!

File: recipes/ocgis/run_test.py (hunk 1 above)
Comment (pelson) [inline recipes/ocgis/run_test.py]: We are leaking GDAL implementation detail into this recipe, which isn't ideal. Maybe we can come up with a better solution in GDAL?

I've raised an issue in https://github.com/conda-forge/gdal-feedstock/issues/35.

Don't let it hold this PR up though - we can always go back and remove it!

Comment (bekozi) [inline recipes/ocgis/run_test.py]: @ocefpaf, this should be good to go with the ``v1.3.1`` tag!

@pelson, I'm not enamored with the solution either. I'd love to find an alternative.

Comment (ocefpaf) [inline recipes/ocgis/run_test.py]: > @ocefpaf, this should be good to go with the ``v1.3.1`` tag!

Rebased, updated, and waiting for the green.

Thanks @bekozi!

**Comment by ocefpaf** [comment]
`ocgis` is fine. The failures are due to a sequence of updates that were not "digested well" by `gdal`. I am working to get `gdal` (`fiona` and `rasterio`) working again so we can merge this.

**Comment by ocefpaf** [comment]
@bekozi the `fiona` issue was solved and I re-started the CIs, but now we have some unitests that are failing: https://circleci.com/gh/conda-forge/staged-recipes/2322

**Comment by bekozi** [comment]
@ocefpaf it looks like the version in the recipe needs to be incremented to ``1.3.1``. Those failures should be fixed.

**Comment by ocefpaf** [comment]
> @ocefpaf it looks like the version in the recipe needs to be incremented to ``1.3.1``. Those failures should be fixed.

Done! (I will be out of office for a few days. If this is green bug @pelson to merge it)

**Comment by jakirkham** [comment]
Have a nice holiday, @ocefpaf.

**Comment by bekozi** [comment]
Thanks, all! :smile:

**Comment by jakirkham** [comment]
Figured we should drop the WIP. :wink:

**Comment by ocefpaf** [comment]
:tada:

**Pull request merged**